src/rewrites.rs

```diff
@@ -849,27 +1037,519 @@ impl MultiPatterns {
+        .newly_added
+        .iter()
+        .map(|node| node.clone().map_children(|id| runner.egraph.find(id)))
+        .collect();
+    let mut added_node_to_order = HashMap::<Mdl, usize>::new();
+    for (i, node) in updated.iter().enumerate() {
+        added_node_to_order.entry(node.clone()).or_insert(i);
+    }
+    // Remove cycles by adding nodes to blacklist
+    remove_cycles(&mut runner.egraph, &added_node_to_order, runner.roots[0]);
+}
+
+/// Add newly added nodes in this pattern to the list of newly added nodes, for use in cycle
+/// filtering
+///
+/// The newly added nodes are stored in the graph level metadata in egraph.analysis
+///
+/// # Parameters
+///
+/// - `pat`: the AST representation of the pattern. See egg::Pattern for more info
+/// - `egraph`: E-graph of interest
+/// - `subst`: mapping variable to eclass ID. See egg::Subst for more info.
+/// - `existing_nodes`: the set of nodes within this pattern that already exists before this
+///         pattern is applied
+///
+/// # Returns
+///
+/// A tuple of (HashSet<Mdl>, Id) where
+///
+/// - HashSet<Mdl>: The set of all nodes in this pattern
+/// - Id: the Id into egraph of the matched root of this pattern
+fn add_newly_added(
+    pat: &[ENodeOrVar<Mdl>],
+    egraph: &mut EGraph<Mdl, TensorAnalysis>,
+    subst: &Subst,
+    existing_nodes: &HashSet<Mdl>,
+) -> (HashSet<Mdl>, Id) {
+    match pat.last().unwrap() {
+        ENodeOrVar::Var(w) => (HashSet::<Mdl>::new(), subst[*w]),
+        ENodeOrVar::ENode(e) => {
+            let children = e.children();
+            let results: Vec<(HashSet<Mdl>, Id)> = children
+                .iter()
+                .map(|child| {
+                    add_newly_added(
+                        &pat[..usize::from(*child) + 1],
+                        egraph,
+                        subst,
+                        existing_nodes,
+                    )
+                })
+                .collect();
+
+            let mut new_e = e.clone();
+            let new_e_ch = new_e.children_mut();
+            for (i, res) in results.iter().enumerate() {
+                new_e_ch[i] = res.1;
+            }
+
+            let mut nodes_in_pat = HashSet::<Mdl>::new();
+            for res in results.iter() {
+                for node in res.0.iter() {
+                    nodes_in_pat.insert(node.clone());
+                }
+            }
+            nodes_in_pat.insert(new_e.clone());
+
+            // Add to order list
+            if !existing_nodes.contains(&new_e) {
+                egraph.analysis.newly_added.push(new_e.clone());
+            }
+
+            (nodes_in_pat, egraph.lookup(new_e).unwrap())
+        }
+    }
+}
+
+/// Remove cycles in EGraph by adding nodes to blacklist
+///
+/// This function works by:
+///     - Make a pass over egraph to get a set of cycles
+///     - For each cycle, pick the node that got added latest and add it to blacklist
+///     - Repeat until no cycles are left
+fn remove_cycles(
```

Comment:
'remove_cycles' is misleading. It doesn't really remove cycles in this function, merely adding nodes to blacklist.

Reply:
Renamed